File: benchmarks/Java/src/som/IdentityDictionary.java
package som;

import som.Dictionary.CustomHash;


public class IdentityDictionary<K extends CustomHash, V> extends Dictionary<K, V> {

  static class IdEntry<K, V> extends Entry<K, V> {
    IdEntry(final int hash, final K key, final V value, final Entry<K, V> next) {
      super(hash, key, value, next);
    }

    @Override
    boolean match(final int hash, final K key) {
      return this.hash == hash && this.key == key;
    }
  }

  public IdentityDictionary(final int size) {
    super(size);
  }

  public IdentityDictionary() {
    super(INITIAL_CAPACITY);
  }

  @Override
  protected Entry<K, V> newEntry(final K key, final V value, final int hash) {
    return new IdEntry<>(hash, key, value, null);
  }
}

File: benchmarks/Python/som/identity_dictionary.py
from som.dictionary import Entry, Dictionary


class _IdEntry(Entry):
    def match(self, hash_, key):
        return self.hash == hash_ and self.key is key


class IdentityDictionary(Dictionary):
    def _new_entry(self, key, value, hash_):
        return _IdEntry(hash_, key, value, None)
